_posts/2019-12-10-bert-model.md

```diff
@@ -0,0 +1,410 @@
+will create an iterator like object for our data.
+
+```jl
+using DelimitedFiles
+
+function gen_train()
+    sets = readdlm("new_pre_train.tsv", '\t', String, quotes = false)
+    return [selectdim(sets, 2, i) for i = (1,2)]
+end
+
+datas = gen_train() # This will be used in the `train!()` function
+```
+
+```jl
+const vocab = Vocabulary(wordpiece)
+
+markline(sent) = ["[CLS]"; sent; "[SEP]"]
+
+function preprocess(batch)
+    sentence = markline.(wordpiece.(tokenizer.(batch[1])))
+    mask = getmask(sentence)
+    tok = vocab(sentence)
+    segment = fill!(similar(tok), 1)
+    
+    label = onehotbatch(batch[2], labels)
+    return (tok = tok, segment = segment), label, mask
+end
+
+preprocess(get_batch(datas, 4))
+```
+
+Here basically what we do it get mask for a sentence, on a way that we don't
+want the model to return the same output. Masks lets us hide it to get a
+different output. We'd be defining the `get_batch` function as well as the
+datas. Here, datas represent the training or testing data we will be using
+in this model.
+
+```jl
+function get_batch(c::Channel, n=1)
+    res = Vector(undef, n)
+    for (i, x) ∈ enumerate(c)
+        res[i] = x
+        i >= n && break
+    end
+    isassigned(res, n) ? batched(res) : nothing
+end
+```
+
+### Step 3: Training and testing the model
+
+First, lets define the `bert_model` and the `clf`.
+
+```jl
+const clf = gpu(Chain(
+        Dropout(0.1),
+        Dense(size(_bert_model.classifier.pooler.W, 1), length(labels)),
+        logsoftmax
+))
+
+const bert_model = gpu(
+    set_classifier(_bert_model,
+                    (
+                        pooler = _bert_model.classifier.pooler,
+                        clf = clf
+                    )
+                  )
+)
+```
+
+We'd be calculating the log cross entropy loss through the loss function:
+
+```jl
+function loss(data, label, mask=nothing)
+    e = bert_model.embed(data)
+    t = bert_model.transformers(e, mask)
+
+    p = bert_model.classifier.clf(
+        bert_model.classifier.pooler(
+            t[:,1,:]
+        )
+    )
+
+    l = Basic.logcrossentropy(label, p)
+    return l, p
+end
+```
+
+Finally, lets define the `train!()` fucntion
+
+```jl
+function train!()
+    global Batch
+    global Epoch
+    @info "start training:"
+    for e = 1:Epoch
+        @info "epoch: $e"
+        datas = datas_tr # Training data generated
+
+        i = 1
+        al::Float64 = 0.
+        while (batch = get_batch(datas, Batch)) !== nothing
+            data, label, mask = todevice(preprocess(batch))
+            l, p = loss(data, label, mask)
+            # @show l
+            a = acc(p, label)
+            al += a
+            grad = gradient(()->l, ps)
+            i+=1
+            update!(opt, ps, grad)
+            i%16==0 && @show al/i
+        end
+
+        test()
+    end
+end
+```
+
+After looping through all of the data, we will move towards testing it:
```

Comment:
This part will make more sense if you've explained above what label you are classifying. Once you have that, maybe refer back to that.